### PR TITLE
Add Perfectionist & Most Creative badges

### PR DESCRIPTION
**Pull request by dillonhammer**
Adds 2 badges based on guess count:
- Perfectionist - No incorrect guesses, minimum 5 words
- Most Creative - Most incorrect guesses, minimum 4 other users

Requires corresponding server PR: https://github.com/dillonhammer/spelling-bee-leaderboard-server/pull/3

Also makes the Changelog collapsible so now we can keep older messages without filling up the homepage.

Screenshot 2023-11-11 at 11 01 16 PM

Screenshot 2023-11-11 at 11 01 58 PM

Screenshot 2023-11-11 at 11 02 21 PM

Screenshot 2023-11-11 at 11 04 43 PM

Screenshot 2023-11-11 at 11 04 48 PM

Adds a `guessCount` field on the user. Initially I was going to have a separate endpoint on the server just for incrementing guess count and have the client fire off that request on every guess, and only update the whole user for correct guesses, but it got annoying to deal with edge cases like if a user's first guess is wrong, then we have to create the whole new user object. So now every single guess fires off an update user request, and there's just a sequence of checks to edit the user payload before it gets sent off.

I think there is technically a race condition possible if a user fires off 2 guesses within milliseconds, since the second one might not know about the result of the first yet and might override it, but looking at the network reqs they seem to return in < 100 ms so I think it'd be very rare to hit that issue. And even if they did it's easy to recover from, just have to enter your word again.

### CODE REVIEW

File: src/components/Badges.tsx [hunk 1]
```diff
@@ -257,6 +325,36 @@ export const generateBadgesForYesterday = (
     );
   }
 
+  if (user.guessCount === user.answers.length && user.answers.length >= 5) {
```
Comment (brianyyang) [inline src/components/Badges.tsx:328]: There's a lot of duplicated code for generating badges for today + yesterday, is it possible to abstract into a helper

File: src/components/HomePage.tsx [hunk 2]
```diff
@@ -132,26 +117,34 @@ const HomePage = ({ userId, navigate }: HomePageProps) => {
...
-    publish("Guess", lowerCaseGuess);
-    if (
-      allUsers &&
-      puzzleData?.answers.includes(lowerCaseGuess) &&
-      !revealedWords.includes(lowerCaseGuess)
-    ) {
-      const sortedWords = [lowerCaseGuess, ...revealedWords].sort((a, b) =>
-        a.localeCompare(b)
-      );
+    if (allUsers && puzzleData && guess.length > 0) {
```
Comment (brianyyang) [inline src/components/HomePage.tsx:120]: More of a game design question, but should we limit the most creative badge for valid guesses? It's another honor system moment where it's like if someone spams "a" as a guess they are kinda lame but should we protect against that.

**Comment by brianyyang** [comment]
Nice badge icons and I like the changelog changes

File: src/components/HomePage.tsx (hunk 2 above)
Comment (dillonhammer) [reply src/components/HomePage.tsx:120]: Yeah, I think it depends on whether people view the badge as something they want to get or do not want to get. I'm not fully sure yet which it'll be. For that example where someone spams, I could see the value in a check where it has to be >= 4 letters and all valid letters in order to count, but I feel like someone who really wants to spam could just spam with that anyway. I'm down to add that check here, I think my hope in general is that at first maybe a few people like to spam but pretty soon after it just becomes like another small piece in the background that is interesting to see but nobody cares enough to grind for it.

File: src/components/Badges.tsx (hunk 1 above)
Comment (dillonhammer) [reply src/components/Badges.tsx:328]: Discussed, will add in a separate PR along with adding badges modals on the yesterday leaderboard.

**Pull request merged**